# libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
from pathlib import Path
import pickle
import json

from statsmodels.tsa.statespace.structural import UnobservedComponents
from statsmodels.tsa.statespace.sarimax import SARIMAX

import sys
sys.path.append('../')
from src import forecasting as fc
print('forecasting imported successfully!')

forecasting imported successfully!


# load data

- "processed" folder containing train and test sets:
  - "C:\Users\aplor\Documents\GitHub\greenhouse-gas-forecasting\data\processed"

- "forecasting_data" folder containing model specs, fitted models, and calibration files:
  - "C:\Users\aplor\Documents\GitHub\greenhouse-gas-forecasting\data\forecasting_data"

## train and test sets

In [2]:
# load the pickle files

output_dir = Path("../data/processed")

train_preprocessed_df = pd.read_pickle(output_dir / "CH4_train_preprocessed.pkl")
test_preprocessed_df = pd.read_pickle(output_dir / "CH4_test_preprocessed.pkl")

In [3]:
train_preprocessed_df.head()

,CH4_preprocessed,value_unc,qcflag
1991-01-06,1719.845,2.71,...
1991-01-13,1736.940,2.71,...
1991-01-20,1728.335,2.71,...
1991-01-27,1716.080,2.71,...
1991-02-03,1727.430,2.71,...


In [4]:
train_preprocessed_df.describe()

,CH4_preprocessed,value_unc
count,1669.000000,1639.000000
mean,1810.978201,1.539019
std,51.454037,0.711564
min,1709.020000,0.575000
25%,1773.175000,0.910000
50%,1799.530000,1.400000
75%,1845.730000,2.050000
max,1970.875000,2.710000


In [5]:
print("train_preprocessed min and max date index:")
print(f"min: {train_preprocessed_df.index.min()}")
print(f"max: {train_preprocessed_df.index.max()}")

train_preprocessed min and max date index:
min: 1991-01-06 00:00:00
max: 2022-12-25 00:00:00


In [6]:
test_preprocessed_df.head()

,CH4_preprocessed,value_unc,qcflag
2023-01-01,1951.4675,0.575,...
2023-01-08,1931.1000,0.575,...
2023-01-15,1933.3625,0.575,...
2023-01-22,1929.8025,0.575,...
2023-01-29,1941.1150,0.575,...


In [7]:
test_preprocessed_df.describe()

,CH4_preprocessed,value_unc
count,105.000000,8.800000e+01
mean,1940.870211,5.750000e-01
std,18.446392,1.116585e-16
min,1906.600000,5.750000e-01
25%,1929.217500,5.750000e-01
50%,1940.986667,5.750000e-01
75%,1952.870000,5.750000e-01
max,1989.377500,5.750000e-01


In [8]:
print("test_preprocessed min and max date index:")
print(f"min: {test_preprocessed_df.index.min()}")
print(f"max: {test_preprocessed_df.index.max()}")

test_preprocessed min and max date index:
min: 2023-01-01 00:00:00
max: 2024-12-29 00:00:00


In [9]:
# confirm that the gap between train and test sets is one week

print(f"last train_preprocessed_df index date: {train_preprocessed_df.index[-1]}")
print(f"first test_preprocessed_df index date: {test_preprocessed_df.index[0]}")

train_test_gap = test_preprocessed_df.index[0] - train_preprocessed_df.index[-1]
print(f"gap between train and test sets = {train_test_gap}")

last train_preprocessed_df index date: 2022-12-25 00:00:00
first test_preprocessed_df index date: 2023-01-01 00:00:00
gap between train and test sets = 7 days 00:00:00


In [10]:
# extract only the CH4 values column from each dataframe

train_preprocessed = train_preprocessed_df['CH4_preprocessed']
test_preprocessed = test_preprocessed_df['CH4_preprocessed']

In [11]:
print(train_preprocessed.describe)

<bound method NDFrame.describe of 1991-01-06    1719.8450
1991-01-13    1736.9400
1991-01-20    1728.3350
1991-01-27    1716.0800
1991-02-03    1727.4300
                ...    
2022-11-27    1958.5475
2022-12-04    1956.4500
2022-12-11    1954.3525
2022-12-18    1952.2550
2022-12-25    1950.1575
Freq: W-SUN, Name: CH4_preprocessed, Length: 1669, dtype: float64>


In [12]:
print(test_preprocessed.describe)

<bound method NDFrame.describe of 2023-01-01    1951.46750
2023-01-08    1931.10000
2023-01-15    1933.36250
2023-01-22    1929.80250
2023-01-29    1941.11500
                 ...    
2024-12-01    1970.82500
2024-12-08    1959.45000
2024-12-15    1941.22625
2024-12-22    1923.00250
2024-12-29    1952.38000
Freq: W-SUN, Name: CH4_preprocessed, Length: 105, dtype: float64>


## model specs, fitted models, and calibration files

In [13]:
# print the path to where the notebook is running
print(Path.cwd())

C:\Users\aplor\Documents\GitHub\greenhouse-gas-forecasting\notebooks


In [14]:
output_dir = Path("../data/forecasting_data")

models_dir = output_dir / "fitted_models"
calibration_dir = output_dir / "calibration"

In [15]:
print(output_dir.resolve())
print(output_dir.exists())

C:\Users\aplor\Documents\GitHub\greenhouse-gas-forecasting\data\forecasting_data
True


**Note about path**
Both data and notebook folders are in the greenhouse-gas-forecasting folder.  

In [16]:
# load the model specs
with open(output_dir / "model_specs.json") as f:
    model_specs = json.load(f)

print("loaded model specs for UCM and SARIMA models")
print(f"\nUCM model specs: {model_specs['UCM']}")
print(f"\nSARIMA model specs: {model_specs['SARIMA']}")

loaded model specs for UCM and SARIMA models

UCM model specs: {'level': 'random walk with drift', 'freq_seasonal': [{'period': 52.18, 'harmonics': 4}], 'stochastic_freq_seasonal': [True], 'autoregressive': 1}

SARIMA model specs: {'order': [1, 1, 1], 'seasonal_order': [1, 0, 1, 52], 'trend': 'c'}


In [17]:
# load fitted models
# I will load only SARIMA and UCM models since these are the top models based on skill and test forecast CRPS 
fitted_models = {}

for name in ["SARIMA", "UCM"]:
    file = models_dir / f"{name}_fit.pkl"
    with open(file, "rb") as f:
        fitted_models[name] = pickle.load(f)

print(f"loaded {len(fitted_models)} fitted models:")
print(list(fitted_models.keys()))

loaded 2 fitted models:
['SARIMA', 'UCM']


In [18]:
# load calibration data for the top models 
# I will load the calibration data for only SARIMA and UCM models since these are the top models based on skill and test forecast CRPS 
calibration = {}

for name in ["SARIMA", "UCM"]:
    file = calibration_dir / f"{name}_calibration.pkl"
    with open(file, "rb") as f:
        calibration[name] = pickle.load(f)

print(f"loaded {len(calibration)} calibration files:")
print(list(calibration.keys()))

loaded 2 calibration files:
['SARIMA', 'UCM']


# single-origin forecast against test set

# forecast against test set - pipeline

1. Fit the model on full training data
2. Run rolling_crps on training data to get the results DataFrame
3. Trim results to common_start_idx=492
4. Apply variance scaling to trimmed results
5. Apply ACI to scaled results, extract final alpha_t
6. Forecast on test set using the fitted model
7. Apply variance scaling and ACI to test forecasts

## single-origin forecast for top models

In [19]:
# Exog for test set per model
# neither model employs an exogenous variable, but am leaving this here for possile
# future use with a model that does.  
exog_test_map = {
    'UCM'    : None,
    'SARIMA' : None
}

all_test_forecasts = {}

for name in fitted_models:
    cal    = calibration[name]        # from run_calibration_pipeline()
    result = fitted_models[name]
    exog_t = exog_test_map[name]

    all_test_forecasts[name] = fc.single_origin_forecast(
        fitted_result=result,
        test_series=test_preprocessed,
        exog_test=exog_t,
        scale_factors=cal['scale_factors'],
        final_alpha=cal['final_alpha'],
        target_coverage=0.95,
    )
    print(f"{name}: forecast complete")

SARIMA: forecast complete
UCM: forecast complete


In [24]:
# Compare all models on the test period
summary_records = []

for name, fc_df in all_test_forecasts.items():
    # Full test period
    summary_records.append({
        'model'        : name,
        'period'       : 'full (2023-2024)',
        'mean_crps'    : fc_df['crps'].mean(),
        'mae'          : fc_df['abs_error'].mean(),
        'coverage_95'  : fc_df['covered_aci'].mean(),
        'n_weeks'      : len(fc_df),
    })
    # Split by year
    for year in [2023, 2024]:
        yr_df = fc_df[fc_df['date'].dt.year == year]
        if len(yr_df) > 0:
            summary_records.append({
                'model'      : name,
                'period'     : str(year),
                'mean_crps'  : yr_df['crps'].mean(),
                'mae'        : yr_df['abs_error'].mean(),
                'coverage_95': yr_df['covered_aci'].mean(),
                'n_weeks'    : len(yr_df),
            })

summary_df = pd.DataFrame(summary_records)
print("Mean CRPS 95% nominal coverage from single-origin forecasts\n")
print(summary_df.pivot(index='model', columns='period',
                        values=['mean_crps', 'coverage_95']).round(3))

Mean CRPS 95% nominal coverage from single-origin forecasts

       mean_crps                          coverage_95                        
period      2023    2024 full (2023-2024)        2023   2024 full (2023-2024)
model                                                                        
SARIMA     8.511  10.028            9.262       0.943  0.981            0.962
UCM        8.079   9.036            8.553       0.981  0.981            0.981


In [21]:
single_origin_test_forecast_summary = pd.DataFrame(summary_df.pivot(index='model', columns='period',
                        values=['mean_crps', 'coverage_95']).round(3))

single_origin_test_forecast_summary

mean_crps                          coverage_95                        
period      2023    2024 full (2023-2024)        2023   2024 full (2023-2024)
model                                                                        
SARIMA     8.511  10.028            9.262       0.943  0.981            0.962
UCM        8.079   9.036            8.553       0.981  0.981            0.981

In [22]:
# general code for including both 80% and 95% nominal coverage inspection

for name, df in aci_test_results_or_similar.items():
    cov_80 = ((df['pit'] >= 0.10) & (df['pit'] <= 0.90)).mean()
    cov_95 = df['covered_aci'].mean()
    print(f"{name}: 80%={cov_80:.3f}, 95%={cov_95:.3f}")

NameError: name 'aci_test_results_or_similar' is not defined

## single-origin forecast comparison plots

In [ ]:
fig, ax = plt.subplots(figsize=(14, 5))

train_tail = train_preprocessed.loc['2021':]
ax.plot(train_tail.index, train_tail.values, color='gray', linewidth=1.2,
        alpha=0.6, label='Training data (2021–2022)')
ax.plot(test_preprocessed.index, test_preprocessed.values, color='black',
        linewidth=1.5, label='Observed (2023–2024)')

colors = {'UCM': '#0072B2', 'SARIMA': '#D55E00'}
for name in ['UCM', 'SARIMA']:
    fc = all_test_forecasts[name]
    ax.plot(fc['date'], fc['y_pred'], color=colors[name],
             linewidth=1.3, linestyle='--', label=f'{name} forecast')
    ax.fill_between(fc['date'], fc['lower_aci'], fc['upper_aci'],
                     color=colors[name], alpha=0.15)

# Seasonal naive reference (52-week lag of training)
naive_vals = train_preprocessed.reindex(
    test_preprocessed.index - pd.Timedelta(weeks=52)
).values
ax.plot(test_preprocessed.index, naive_vals, color='gray',
        linewidth=1, linestyle=':', label='Seasonal naive')

ax.axvline(train_preprocessed.index[-1], color='gray', linestyle=':', linewidth=1)
ax.set_title('CH4 Forecast Comparison: 2023–2024', fontsize=13)
ax.set_ylabel('CH4 (ppb)')
ax.legend(fontsize=9, loc='upper left')
ax.grid(True, linestyle='--', alpha=0.4)
plt.tight_layout()
plt.show()

Suggested summary paragraph for the notebook:

On the held-out 2023–2024 test period, both UCM and SARIMA substantially outperform the seasonal naive benchmark [insert naive CRPS once computed], confirming genuine forecasting skill beyond simple year-over-year persistence. UCM achieves the lowest mean CRPS (8.55 ppb, vs. SARIMA's 9.26 ppb) across the full test period, with both models showing modestly higher error in 2024 than 2023 — consistent with the known anomalous CH4 suppression observed during 2024. Empirical coverage at the nominal 95% level is 0.981 for UCM and 0.962 for SARIMA — both above target, indicating the calibrated prediction intervals are conservative (wider than strictly necessary) rather than overconfident. This is attributable to the ACI calibration inheriting wide interval widths learned during the volatile 2020–2022 training period, a known limitation discussed further in Section [X].

Part B (refit on full data, forecast into 2025)

The 2025 Forecast — Order of Operations:

Your seven-step list is close, but has a redundancy worth removing: you should not re-run steps 2–5 for the 2025 forecast. Those steps (rolling CRPS on training data → trim → variance scaling → ACI → extract final alpha_t) are exactly what you already did to produce calibration['UCM'] and calibration['SARIMA'], which you've already saved and loaded. Recomputing them for 2025 would be pure duplication of work you've already validated — and would also not actually change anything, since they don't depend on the forecast horizon target, only on the training data (which hasn't changed).

The corrected, non-redundant order for the 2025 forecast is:

1. Refit the model on the combined train+test data (all observed data through Dec 2024) — this is new, since your current fitted_models were fit on train only.
2. Reuse the existing calibration['UCM'] (scale_factors, final_alpha) — already computed, already saved, no need to redo.
3. Generate the forecast for the desired 2025 horizon using the refit model (get_forecast(steps=52) or however far into 2025 you want).
4. Apply variance scaling and ACI to that forecast using the existing calibration parameters from step 2 — exactly the same single_origin_forecast-style logic you already built and validated on the test set.

In [ ]:
# Step 1 — refit on ALL observed data
full_series = pd.concat([train_preprocessed, test_preprocessed])

final_ucm_2025 = UnobservedComponents(full_series, **model_specs['UCM'])
final_ucm_2025_result = final_ucm_2025.fit(method='lbfgs', maxiter=2000, disp=False)

# Step 2 — reuse existing calibration (already loaded)
cal = calibration['UCM']

# Step 3 & 4 — forecast + calibrate, using your existing single_origin_forecast()
# but there's no test_series to score against yet — you need a lightweight
# variant that just produces mu/lower/upper without requiring y_true

n_forecast = 52  # weeks into 2025
fc_2025 = final_ucm_2025_result.get_forecast(steps=n_forecast)
mu_2025    = fc_2025.predicted_mean.values
sigma_2025 = fc_2025.se_mean.values

sf_horizons    = sorted(cal['scale_factors'].index.tolist())
alpha_horizons = sorted(cal['final_alpha'].index.tolist())

rows = []
for i in range(n_forecast):
    h = i + 1
    h_sf    = max([x for x in sf_horizons if x <= h], default=sf_horizons[0])
    h_alpha = max([x for x in alpha_horizons if x <= h], default=alpha_horizons[0])
    sigma_cal = sigma_2025[i] * cal['scale_factors'].loc[h_sf]
    z_t = stats.norm.ppf(1 - cal['final_alpha'].loc[h_alpha] / 2)
    rows.append({
        'date'  : mu_2025... # build date index from full_series.index[-1] + i+1 weeks
        'y_pred': mu_2025[i],
        'lower' : mu_2025[i] - z_t * sigma_cal,
        'upper' : mu_2025[i] + z_t * sigma_cal,
    })
forecast_2025_df = pd.DataFrame(rows)

In [ ]:
# combine the preprocessed train and test sets
full_series = pd.concat([train_preprocessed, test_preprocessed])

print(f"full_series first and last values at: {full_series.index.date[0]} and {full_series.index.date[-1]}")

In [ ]:
# Fit the saved UCM model_specs on the preprocessed full CH4 series 

full_series_ucm = UnobservedComponents(
    full_series,
    **model_specs['UCM']
)

full_series_ucm_results = full_series_ucm.fit(
    method='lbfgs',
    maxiter=2000,
    disp=False
)